### ISSUE TITLE
regression: NIRCam/FGS SFLAT/DFLAT/FFLAT new keywords

### ISSUE THREAD

**Issue by stscijgbot** [spacetelescope/jwst#3935]
Issue [JP-944](https://jira.stsci.edu/browse/JP-944) was created by Jonathan Eisenhamer:

Two tests have failed, all related to DFLAT/FFLAT/SFLAT new keywords. Confirm and update truth files.

 
 * [jwst.tests_nightly.general.fgs.test_guider_pipeline.TestGuiderPipeline.[stable-deps] test_guider_pipeline2|https://plwishmaster.stsci.edu:8081/job/RT/job/JWST/357/testReport/junit/jwst.tests_nightly.general.fgs.test_guider_pipeline/TestGuiderPipeline/_stable_deps__test_guider_pipeline2/]
 * [jwst.tests_nightly.general.nircam.test_image2pipeline_2b.TestImage2Pipeline.[stable-deps] test_image2pipeline2b|https://plwishmaster.stsci.edu:8081/job/RT/job/JWST/357/testReport/junit/jwst.tests_nightly.general.nircam.test_image2pipeline_2b/TestImage2Pipeline/_stable_deps__test_image2pipeline2b/]

**Comment by stscijgbot** [comment]
Comment by Nadia Dencheva: After a long time digging through the Jenkins logs I found that the syntax with hashes in setup.py actually doesn't work as expected. Although the logs say astropy is collected at the specific hash, what gets installed is astropy master. This can explain why some tests started failing without a change in the repo.

**Comment by stscijgbot** [comment]
Comment by Steven Crawford: In the last few regression tests this seems still to be a problem.  In tests 369, 370, 371, and 372, there are different tests that fail with this error.  The test will pass in the next version and then a different test will fail.   As of 371, there were no failures that was not related to this problem.

**Comment by stscijgbot** [comment]
Comment by Steven Crawford: Conversation with [~hodge] [~jmiller] with some more information:

 

    I've been out of the loop for a week or so,  can someone forward me any

    new status?

   

    On Sep 4 2019 2:02 PM, Philip Hodge wrote:

    > It's never R_FLAT that's missing, just the other three.  One thing

    > that's unusual about the flat_field step is that it has more than one

    > reference file; that's not unique, but I'm not aware of any other step

    > that has four reference types.

    I thought FLAT was always there and the _ versions faded in and out.

    > The fact that the error is not consistent suggests a timing issue.

    > Would it be practical to insert a call to time.sleep() somewhere in

    > the prefetch stage of CRDS, maybe sleep for one second?

    Sure,  but between what and what?   Is any multiprocessing going on in

    CAL?   (We could also do this test outside Jenkins to reproduce Jenkins)

    > Not permanently, just as a test.  Another possible test that would

    > have less of an impact on everybody would be to explicitly specify the

    > dflat, fflat, and sflat files for one or two regression tests,

    > preferably calwebb_spec2 tests because that's where we have seen the

    > most failures; that is, specify the paths to the files that CRDS would

    > have found anyway, just to bypass the prefetch step. If those tests

    > then pass, that would suggest the problem has to do with the prefetch.

   

    What does seem clear is that STPIPE is somehow producing the wrong

    prefetch list.   So CRDS per-say is not relevant,  just the way in which

    jwst.stpipe.pipeline.Pipeline is constructing the prefetch list.   That

    should only depend on the sequence of steps in the pipeline.

   

    The corrupt association file hasn't been resolved either.   It makes me

    wonder if Pipeline._is_container() is working.  If that fails on a

    container,  Pipeline may prefetch incorrectly.

**Comment by jdavies-st** [comment]
I do recall that in the past when running CRDS using `pytest -n N`, i.e. parallelizing it, if we were running outside of the ST network so that CRDS needed to actually make requests for the files, we needed to set filelock.  For people who do not have access to central store, this is required.

Running inside the network, the cache on central store is used.  But perhaps there's still a parallelization issue?  @jaytmiller ?

**Comment by jaytmiller** [comment]
Answering your question @jdavies,  parallelization could certainly still 
be an issue,  but:

1. CRDS file locking is only used to protect the CRDS cache from 
concurrent writes which I don't think are happening here.

2. This looks like it's happening before CRDS is called, somewhere in 
stpipe while computing what to prefetch or get.

3. I may be reading this wrong,  but it looks like Jenkins is not using 
pytest -n and some failing tests look concurrency-free.

*More importantly,  I may have figured this out!*

It's a quasi-timing issue:  non-deterministic test ordering. (There's a 
package called pytest-ordering so I'm guessing the pytest default is 
variably ordered.)

This code in flatfield/tests/test_flatfield.py is mutating the /class/ 
from which other tests are later constructing FlatFieldSteps:

    def test_flatfield_step_interface(instrument, exptype):
         """Test that the basic inferface works for data requiring a
    FLAT reffile"""
         ...

         # override class attribute so only the `flat` type needs to be
    overriden
         # in the step call.  Otherwise CRDS calls will be made for the
    other 3
         # types of flat reference file not used in this test.

         FlatFieldStep.reference_file_types = ["flat"]      # <---  ding
    ding ding

         result = FlatFieldStep.call(data, override_flat=flat)

         assert (result.data == data.data).all()
         assert result.var_flat.shape == shape
         assert result.meta.cal_step.flat_field == 'COMPLETE'

After the above test runs,  the FlatFieldStep class may be broken,  it's 
a global side effect.

Other tests running afterward will likewise omit DFLAT, FFLAT, SFLAT.

Comments anyone?

Todd

**Comment by stscijgbot** [comment]
Comment by Todd Miller: I'm now pretty sure that flatfield/tests/test_flatfield.py is the problem.   We could disable the test,  save and restore the reference_file_types,  or maybe make a disposable deepcopy of FlatFieldStep and modify that for the test.   Just saving and restoring would be a latent issue awaiting the day we do start using pytest -n.

**Comment by stscijgbot** [comment]
Comment by Philip Hodge: Would there be any concurrency issues with making a deepcopy of FlatFieldStep and modifying that?  I don't see why there wouldn't be, but if not, that would be a good solution.  Otherwise, I would suggest disabling the test; the step interface will get tested by regression tests anyway.

**Comment by stscijgbot** [comment]
Comment by Steven Crawford: Nice catch [~jmiller] !   These errors appeared again last night, so it would be good to disable the test for the time being.  [~hodge] can you go ahead and submit a PR to disable it for the time being?

**Comment by stscijgbot** [comment]
Comment by Todd Miller: It dawned on me that hacking reference_file_types might not be a good test after all.  It needs to work.

**Comment by stscijgbot** [comment]
Comment by Steven Crawford: The error hasn't re-appeared so closing this as removing this test seems to have worked

**Comment by stscijgbot** [comment]
Comment by Steven Crawford: For the record, the test is located in this file:

[https://github.com/spacetelescope/jwst/blob/master/jwst/flatfield/tests/test_flatfield.py]

**Comment by stscijgbot** [comment]
Comment by Nadia Dencheva: I think modifying a class variable is rarely a good idea. We should probably remove the test (it should not be a valid use case), or modify it to use "override".

Good catch [~jmiller]!

**Comment by jdavies-st** [comment]
Thanks for the sleuthing Todd.  Yeah, I wrote that test and did not think of the side-effects of modifying a class attribute.

This didn't manifest itself earlier this year when I wrote it, as the unit tests were not being run with the regression tests at any point.  But once we modified the JenkinsfileRT to run all the tests nightly (unit + regression) that's when it showed up.

And these tests are being parallelized `pytest -n 30` across 30 different cores.  I suspect that contributes to the non-repeatable nature of this problem, and also why if just the regression tests themselves are run, the problem doesn't show up.

What would be the better way to run this test, or a modified test?  Clearly mutating the class attribute is not good, or at least leaving it mutated is not good.
